File: java-data-processing-system/src/Worker.java
import java.util.concurrent.ThreadLocalRandom;

// Worker class using Thread execution
public class Worker extends Thread {
    private final TaskQueue taskQueue;
    private final int workerID;

    public Worker(TaskQueue taskQueue, int workerID) {
        this.taskQueue = taskQueue;
        this.workerID = workerID;
    }

    @Override
    public void run() {
        System.out.println("Worker " + workerID + " started.");
        while (true) {
            try {
                Task task = taskQueue.getTask();
                // Check for termination signal (null task)
                if (task == null) {
                    System.out.println("Worker " + workerID + " received termination signal.");
                    break;
                }

                // Process the task
                int taskID = task.getTaskID();
                String taskData = task.getData();
                System.out.println("Worker " + workerID + " processing task: " + taskID + " with data: " + taskData);

                // Simulate processing delay (some kind of computational work)
                Thread.sleep(ThreadLocalRandom.current().nextInt(500, 1000));

                // Create result and write it to the shared output file
                String result = "Worker " + workerID + " completed task: " + taskID + " with data: " + taskData;
                ResultWriter.writeResult(result);
                System.out.println("Worker " + workerID + " finished task: " + taskID + " with data: " + taskData);
            } catch (InterruptedException e) {
                String result = "Worker " + workerID + " was interrupted: " + e.getMessage();
                ResultWriter.writeResult(result);
                System.err.println(result);
                break;
            } catch (Exception e) {
                String result = "Worker " + workerID + " encountered error: " + e.getMessage();
                ResultWriter.writeResult(result);
                System.err.println(result);
            }
        }
        System.out.println("--- Worker " + workerID + " terminated. ---");
    }
}

File: go-data-processing-system/worker.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"time"
)

// worker function processes tasks from the 'tasks' channel and
// sends results/errors to the 'results' channel.
func worker(workerID int, tasks <-chan *Task, results chan<- string) {
	log.Printf("Worker %d started.\n", workerID)

	// Range over the tasks channel. This loop automatically exits
	// when the tasks channel is closed and drained.
	for task := range tasks {
		// Check for termination signal (nil task).
		if task == nil {
			log.Printf("Worker %d received termination signal.\n", workerID)
			break
		}

		log.Printf("Worker %d processing task %d: %s\n", workerID, task.TaskID, task.Data)

		// Simulate computational work with a delay
		processingTime := time.Duration(rand.Intn(500)+500) * time.Millisecond
		time.Sleep(processingTime)

		result := fmt.Sprintf("Worker %d completed task %d: %s", workerID, task.TaskID, task.Data)
		results <- result
		log.Printf("Worker %d finished task %d: %s\n", workerID, task.TaskID, task.Data)
	}
	log.Printf("Worker %d terminated.\n", workerID)
}
